include/envoy/init/init.h

```diff
@@ -55,9 +38,24 @@ class Manager {
   };
 
   /**
-   * Returns the current state of the init manager.
+   * @return the current state of the manager.
    */
   virtual State state() const PURE;
+
+  /**
+   * Register an initialization target. If the manager's current state is uninitialized, the target
+   * will be saved for invocation later, when `initialize` is called. If the current state is
+   * initializing, the target will be invoked immediately. It is an error to register a target with
+   * a manager that is already in initialized state.
+   * @param target_receiver the target to be invoked when initialization begins.
+   */
+  virtual void add(const TargetReceiver& target_receiver) PURE;
```

Comment:
Is there a constraint that Receiver functors can't have any by-reference capture? Is there a way to enforce that at the type-system level(if they were plain-old C function pointers, that would do that..)?

Reply:
So, yeah, ideally you'd want something like Rust's borrow checker to enforce that your callbacks don't capture something that is destroyed before the callback is invoked. But there doesn't appear to be any such mechanism. So the basic idea behind this design is that the `Receiver` has the same (or narrower) lifetime as any references captured by its callback, which is pretty easy to get right and takes some effort to get wrong.